8 register user

jublebi opened the pull request:
### What is done:
- Added initial structure for the user service. 
- In this case, the user has a unique identifier ID, email, and username. 
- In addition, the user must have a first and last name. 
- Optional a user can have an address object that could be used for search options later. Since a user might want multiple addresses, the address is a separate entity.
- Database runs in a docker container on my device.

### known issues
- The tests cover only the storing of the user and address in the database 
- UserService only searches for uuid. We must consider methods to deal with the three different unique identifiers. We could use the ID on the server, but the client can also send an email or username, which gets mapped to UUID on the server. 
- User has no password. We should discuss how we wanna add this aspect.

### For the reviewer:
Please check the following: 
- [ ]  If the structure is okay for the use case.
- [ ] Check if the entities user and address make sense
- [ ] If the API routes are okay

Commit:
Switched to docker

Instead of local postgres installation, docker is used to create container for it.

Commit:
Fixed Error table not present

Used GeneratedValue with Identity as strategy. But it is not allowed for UUID. Switched to Type Auto and it works.

Commit:
Added missing classes for address

Now everything which is needed to access the address via the api is implemented.

Commit:
Refactored Address Service Test

Moved it to the actual service as not the ressource is tested.

Commit:
Tested create user within UserService

tested if a user object can be stored into database correctly.

Commit:
Added addAddressToUser method

now an address can be added to a given user ID.

Commit:
Added lombok

To make getter and setter with annotation.

Commit:
Tested add address to user

Commit:
Removed unused imports

Commit:
Added username and email to user

Now two additional unique parameters exist as we want a login via email or username. But we have to consider how we deal with 3 unique identifiers.

Commit:
Initial commit on feature branch

rmzzz commented on pom.xml:
```diff
@@ -107,6 +107,17 @@
         </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
```
we don't need lombok, because we don't actually need getters/setters in the entities.

rmzzz commented on src/main/java/at/aau/ase/cl/api/AddressResource.java:
```diff
@@ -0,0 +1,46 @@
+    @Path("address")
+    public Response createAddress(Address address) {
+        // create address
+        System.out.println("Request angekommen");
```
please avoid stdout, use `Log.debug()` or `Log.trace()` instead

rmzzz commented on src/main/java/at/aau/ase/cl/api/AddressResource.java:
```diff
@@ -0,0 +1,46 @@
+        // create address
+        System.out.println("Request angekommen");
+        var model = AddressMapper.INSTANCE.map(address);
+        System.out.println("model gemappt");
```
let's write the whole code in english ;)

rmzzz commented on src/main/java/at/aau/ase/cl/api/model/Address.java:
```diff
@@ -2,9 +2,11 @@
 
 import java.util.UUID;
-
+@Getter @Setter
```
we don't need getters and setters in data classes. In book service e.g. I have replaced DTO classes with records, you can user them too if you like.

rmzzz commented on src/main/java/at/aau/ase/cl/api/model/Address.java:
```diff
@@ -2,9 +2,11 @@
 import java.util.UUID;
-
+@Getter @Setter
 public class Address {
```
we should use countryCode instead of full country name. see https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2
in Java we can get the actual country codes programmatically from `Locale` class

rmzzz commented on src/main/java/at/aau/ase/cl/model/AddressEntity.java:
```diff
@@ -1,18 +1,23 @@
 import java.util.UUID;
 
 @Entity
 @Table(name = "address")
```
lets use table prefix for the service, e.g. "user_address", in this way we could use single database for all microservices without naming conflicts

rmzzz commented on src/main/java/at/aau/ase/cl/model/UserEntity.java:
```diff
@@ -11,6 +11,7 @@
 @Table(name = "users") //user table may be already in use
```
its easier to keep table name in singular, i.e. "user"

rmzzz commented on src/main/java/at/aau/ase/cl/service/UserService.java:
```diff
@@ -14,7 +15,9 @@ public class UserService {
         // create user
-        user.id = UUID.randomUUID();
+        if (UserEntity.find("email", user.email).firstResult() != null) {
+            throw new IllegalArgumentException("A user with this email already exists: " + user.email);
```
unless you provide an extra ExceptionMapper, this will result in internal server exception (500), which is not quite REST conform, because it is client error, not server side. That is, we could return conflict (409) or bad request (400). The later can be used directly as BadRequestException, the conflict is not provided, but we can use ClientErrorException with providing explicit status as parameter.

rmzzz commented on src/main/java/at/aau/ase/cl/service/UserService.java:
```diff
@@ -14,7 +15,9 @@ public class UserService {
+        if (UserEntity.find("email", user.email).firstResult() != null) {
+            throw new IllegalArgumentException("A user with this email already exists: " + user.email);
+        }
         user.persist();
```
the check above is not guaranteed collision free. that is, it can happen that concurrent transaction is committed between the check and update. Then the unique constrain violation will be thrown from the hibernate/database layer.
We can catch it too using `user.persistAndFlush()` instead of just persist, surrounded by try-catch block that catches PersistenceException or its subclasses.

rmzzz commented on src/test/java/at/aau/ase/cl/service/UserServiceTest.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    @Test
+    @Transactional
```
the service methods under test are Transactional, why do we need Transaction on the test method?
If it is here to compensate issues related to the lazy loading, then the same issues will be in the REST resource bean that uses these services.
therefore, the tests of the REST resources - UserResource and AddressResource - are needed as well.

rmzzz commented on src/main/java/at/aau/ase/cl/api/AddressResource.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.UUID;
+
+@Path("/")
+public class AddressResource {
```
to have separate address endpoint is ok,
but iff address is one-to-one associated with user it would be probably easier to manage them together, moreover, it is already part of the user object.

rmzzz commented on src/main/java/at/aau/ase/cl/model/UserEntity.java:
```diff
@@ -19,6 +20,12 @@ public class UserEntity extends PanacheEntityBase {
+    @Column(name = "username", nullable = false, unique = true)
+    public String username;
+
     @OneToOne(cascade = CascadeType.ALL, fetch = FetchType.LAZY, optional = true)
```
iff address is associated one-to-one, that is, never reused, and you decide (see other comments) to remove address endpoint and manage user and address always together, we could embed it into the user table.
to achieve this, the AddressEntity can be annotated as `@Embeddable` instead of `@Entity` and `@Table`.

rmzzz commented on src/main/java/at/aau/ase/cl/api/UserResource.java:
```diff
@@ -39,4 +38,18 @@ public Response getUser(@PathParam("id") UUID id) {
+            var result = UserMapper.INSTANCE.map(modelUser);
+            return Response.ok(result).build();
+        } catch (NotFoundException e) {
+            return Response.status(Response.Status.NOT_FOUND).entity("User with ID " + id + " not found").build();
```
this catch will work, but it is not necessary, because `NotFoundException` from jax rs is translated into the not-found status code (404) out of the box.

rmzzz reviewed (changes requested):
The overall implementation is very good for the first time with the new framework.
Please consider few hints and fix the issues found.

Commit:
Changed Logging

Now we are using Log.debug

Commit:
Changed createUser

Now we use persistAndFlush to avoid collisions. We do it within a try catch block. If a ConstraintViolationException occurs, a user with the email already exists.

Commit:
Created IllegalArgumentExceptionMapper

To be rest conform we map the IllegalArgumentException which would result in error code 500 (internal server error) to a bad request exception which maps to the error code 400.

Commit:
Changed Address Entity to Embeddable

Since it is mapped one to one with the user it makes more sense to store it inside the user and not in a separate table.

Commit:
Changed Country to CountryCode

instead of storing the country name we store the country code following the regex "^[A-Z]{2}$" . We also validate if the country code exists using java.util.Locale

Commit:
Changed Country to CountryCode

instead of storing the country name we store the country code following the regex "^[A-Z]{2}$" . We also validate if the country code exists using java.util.Locale

Commit:
Tested User Resource

implemented first test for the user service. Need to create a clean database method so that every test starts with a fresh database.

djamn committed:
add cors access control interceptor
Adds cors access to header in order to access backend via frontend

Commit:
Added Test Cases

Commit:
Changed User Entity

Removed firstname and lastname of user as they are not needed. Instead, username will be used.

Commit:
Added password to user

Now the user entity stores a password hash. The plaintext password which comes in when creating the user gets hashed within the createUser method and stored.

Commit:
Implemented basic Security logic

Added basic authentication and jakarta persistence to enable role based authentication. Now we can permit access to certain resources to certain roles. It is implemented now as the user object had to be adapted accordingly.

Commit:
Started using JWT as login method

not complete but complete to save progress.

Commit:
Implemented basic Security logic

it now works to login and get a JWT which can be validated for a role. Further implementation needs to be done to actually secure paths.

Commit:
Implemented basic Security logic

Now it actually works to validate generated tokens and paths can be secured for given roles. Further changes need to be done as currently the role is transmitted via the request. it actually should happen by the backend.
